### PR TITLE
Geometry, FieldDefinition, GeometryFieldDefinition, CoordinateTransformation auto pointers

### PR DESCRIPTION
**Pull request by turboladen**
Relates to #28, where lots of wrapped pointers in the library don't get cleaned up when GC happens. Geometry is probably the most widely used culprit, but there are definitely a lot more of these to change to follow suit. I didn't check, but I imagine that this should take care of a bunch of memory leaks.

- [x] Update commits with ticket number

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -29,8 +29,9 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
   - `OGR::MultiSurface`
   - `OGR::Surface`
 - Added `.new_from_coordinates` for `OGR::Point` and `OGR::Point25D`.
-- Added `OGR::Geometry#to_gml_ex` (well, changed `#to_gml` to wrap
-  `OGR_G_ExportToGML` instead of `OGR_G_ExportToGMLEx`.
+- Added `OGR::Geometry#to_gml_ex` (well, changed `#to_gml` to wrap `OGR_G_ExportToGML` instead of
+  `OGR_G_ExportToGMLEx`.
+- Added `OGR::FieldDefinition.new_borrowed` for instantiating from borrowed pointers.
```
Comment (turboladen) [inline Changelog.md:34]: Same for OGR::SpatialReference

File: Changelog.md [hunk 2]
```diff
@@ -158,6 +185,10 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
   - `OGR::SpatialReferenceMixins::CoordinateSystemGetterSetters#authority_name`
   - `OGR::StyleTool#param_as_string`
   - `OGR::StyleTool#style_string`
+- #28: Leveraged `FFI::AutoPointer` to ensure wrapped pointers get cleaned up; the following
+  definitely weren't:
+  - any `OGR` geometry
+  - `OGR::CoordinateTransformation`
```
Comment (turboladen) [inline Changelog.md:191]: update this list

**Review by kurtronshausen — APPROVED** [review]

File: lib/ffi/ogr/api.rb [hunk 3]
```diff
@@ -13,253 +16,18 @@ module API
...
       typedef :pointer, :OGRFeatureH
       typedef :pointer, :OGRStyleTableH
-      typedef :pointer, :OGRGeomFieldDefnH
       typedef :pointer, :OGRLayerH
       typedef :pointer, :OGRDataSourceH
       typedef :pointer, :OGRSFDriverH
       typedef :pointer, :OGRStyleMgrH
       typedef :pointer, :OGRStyleToolH
 
-      # -----------------------------------------------------------------------
```
Comment (turboladen) [inline lib/ffi/ogr/api.rb:28]: I moved all of these out to child files simply so I could gradually add type-checking for the functions in this module

File: lib/ffi/ogr/api/geometry.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      attach_function :OGR_G_AddGeometry, %i[OGRGeometryH OGRGeometryH], FFI::OGR::Core::Err
+      attach_function :OGR_G_AddGeometryDirectly, %i[OGRGeometryH OGRGeometryH], FFI::OGR::Core::Err
+      attach_function :OGR_G_RemoveGeometry, %i[OGRGeometryH int bool], FFI::OGR::Core::Err
+
+      attach_function :OGRBuildPolygonFromEdges,
+                      %i[OGRGeometryH bool bool double pointer],
+                      :OGRGeometryH
+      # end
+      # end
+      # end
```
Comment (turboladen) [inline lib/ffi/ogr/api/geometry.rb:199]: nuke

File: lib/ogr/coordinate_transformation.rb [hunk 5]
```diff
@@ -1,44 +1,30 @@
...
+    # @return [OGR::CoordinateTransformation::AutoPointer] C pointer that
+    #   represents the CoordinateTransformation.
     attr_reader :c_pointer
 
     # @param source_srs [OGR::SpatialReference]
     # @param destination_srs [OGR::SpatialReference]
     # @raise [FFI::GDAL::InvalidPointer]
     def initialize(source_srs, destination_srs)
-      source_ptr = GDAL._pointer(source_srs)
-      destination_ptr = GDAL._pointer(destination_srs)
```
Comment (turboladen) [inline lib/ogr/coordinate_transformation.rb:41]: FWIW, I feel like this became an anti-pattern, obfuscating what's going on in that method when we really should be able to determine in each method whether we're working with a borrowed or owned pointer. ...thus I'm getting rid of these calls as I go, where it makes sense.

File: sig-gems/ffi/1.15.3/ffi/library.rbs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    # | (
+      # _ToS ruby_method_name,
+      # _ToS c_function_name,
+      # Array[Symbol | FFI::Enum] args,
+      # ?Symbol returns,
+      # ?blocking: bool,
+      # ?convention: Symbol,
+      # ?enums: FFI::Enums,
+      # ?type_map: Hash
+    # ) -> FFI::VariadicInvoker
```
Comment (turboladen) [inline sig-gems/ffi/1.15.3/ffi/library.rbs:49]: nuke

File: sig-gems/ffi/1.15.3/ffi/library.rbs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        # FFI::DataConverter | FFI::Struct | Symbol | FFI::Type ret
+      # ) -> FFI::CallbackInfo
+
+    # def attach_variable:
+      # (_ToS mname, _ToS cname, FFI::DataConverter | FFI::Struct | Symbol | FFI::Type var_type) -> FFI::DynamicLibrary::Symbol
+    # | (_ToS cname, FFI::DataConverter | FFI::Struct | Symbol | FFI::Type var_type)             -> FFI::DynamicLibrary::Symbol
+
+    # def callback:
+      # (Symbol name, Array[Symbol | FFI::Enum | FFI::Type] params, FFI::DataConverter | FFI::Struct | Symbol | FFI::Type ret) -> FFI::CallbackInfo
+    # | (Array[Symbol | FFI::Enum | FFI::Type] params, FFI::DataConverter | FFI::Struct | Symbol | FFI::Type ret)              -> FFI::CallbackInfo
```
Comment (turboladen) [inline sig-gems/ffi/1.15.3/ffi/library.rbs:76]: nuke

File: sig/ogr/coordinate_transformation.rbs [hunk 8]
```diff
@@ -1,5 +1,53 @@
 module OGR
   class CoordinateTransformation
-    include _HasPointer
+    class AutoPointer < FFI::AutoPointer
+      def self.release: (FFI::Pointer c_pointer) -> void
```
Comment (turboladen) [inline sig/ogr/coordinate_transformation.rbs:4]: use the interface def

**Review by tindron — APPROVED** [review]

**Comment by turboladen** [comment]
gah, accidentally rebased old commits here 🤦

**Pull request merged**